File: HermiteInterpolation/SplineKnots/KnotsGenerator.cs
using HermiteInterpolation.Numerics;
using HermiteInterpolation.Numerics.MathFunctions;
using HermiteInterpolation.Shapes.SplineInterpolation;

namespace HermiteInterpolation.SplineKnots
{
    public abstract class KnotsGenerator
    {
        protected KnotsGenerator(InterpolativeMathFunction function)
        {
            Function = function;
        }

        protected KnotsGenerator()
        {
        }

        protected KnotsGenerator(MathExpression expression)
            : this(InterpolativeMathFunction.FromExpression(expression))
        {
        }

        public InterpolativeMathFunction Function { get; }

        public abstract KnotMatrix GenerateKnots(SurfaceDimension uDimension,
            SurfaceDimension vDimension);

        public static KnotsGenerator operator +(
            KnotsGenerator leftOp, KnotsGenerator rightOp)
        {
            return new ChainedKnotsGenerator(leftOp)
            {
                {rightOp, (l, r) => l + r}
            };
        }

        public static KnotsGenerator operator -(
            KnotsGenerator leftOp, KnotsGenerator rightOp)
        {
            return new ChainedKnotsGenerator(leftOp)
            {
                {rightOp, (l, r) => l - r}
            };
        }
    }
}

File: SplineKnotsBenchmark/KnotsGenerator.cpp
#include "stdafx.h"
#include "KnotsGenerator.h"

namespace splineknots
{
	const InterpolativeMathFunction& KnotsGenerator::Function() const
	{
		return function_;
	}

	KnotsGenerator::KnotsGenerator(MathFunction function)
		: function_(function)
	{
	}

	KnotsGenerator::KnotsGenerator(InterpolativeMathFunction function)
		: function_(function)
	{
	}
}
